core/src/Cabana_Parallel.hpp

```diff
@@ -143,244 +49,92 @@ class StructAndArrayParallelTag {};
-
-  Takes an instance of the \c StructAndArrayParallelTag to indicate specialization.
-
-  Creates a 2D parallel loop over structs and their inner arrays:
-  \code
-  AoSoA aosoa;
-  Index begin, end;
-  parallel_for( s : num_structs )
-  {
-  parallel_for( a : array_size(s) )
-  {
-  Index idx( array_size, s, a );
-  functor( idx );
-  }
-  }
-  \endcode
+  In the above example, \c struct defines an index to a given AoSoA/Slice
+  struct and array defines and index to the given array element in that struct.
+  Its <tt>operator()</tt> method defines the operation to parallelize, over
+  the range of indices <tt>idx=[begin,end]</tt>. The kernel represented by the
+  functor is intended to vectorize of the array index.
 */
-template<class ExecutionPolicy, class FunctorType>
-inline void parallel_for( const ExecutionPolicy& exec_policy,
-                          const FunctorType& functor,
-                          const StructAndArrayParallelTag&,
-                          const std::string& str = "" )
+template<class ExecutionSpace, class FunctorType, int VectorLength>
+inline void parallel_for(
+    const RangePolicy2d<ExecutionSpace,VectorLength>& exec_policy,
+    const FunctorType& functor,
+    const std::string& str = "" )
 {
-    // Type aliases.
-    constexpr auto kokkos_iterate= Kokkos::Iterate::Right;
     using kokkos_policy =
-        Kokkos::MDRangePolicy<typename ExecutionPolicy::execution_space,
-                              Kokkos::Rank<2,kokkos_iterate,kokkos_iterate>,
-                              Kokkos::IndexType<int> >;
-    using point_type = typename kokkos_policy::point_type;
+        Kokkos::TeamPolicy<ExecutionSpace,
+                           Kokkos::IndexType<int>,
+                           Kokkos::Schedule<Kokkos::Dynamic> >;
 
-    // Make a 2D execution policy.
-    auto begin = exec_policy.begin();
-    auto end = exec_policy.end();
-    auto struct_bounds =
-        Impl::getStructBounds<ExecutionPolicy::array_size>( begin, end );
-    point_type lower = { struct_bounds.first, 0 };
-    point_type upper = { struct_bounds.second, ExecutionPolicy::array_size };
-    kokkos_policy k_policy( lower, upper );
+    kokkos_policy team_policy( exec_policy.numStruct(), 1, VectorLength );
 
-    // Create a wrapper for the functor.
-    auto functor_wrapper =
-        KOKKOS_LAMBDA( const std::size_t s, const int a )
+    Kokkos::parallel_for(
+        str,
+        team_policy,
+        KOKKOS_LAMBDA( const typename kokkos_policy::member_type& team )
         {
-            auto array_bounds =
-            Impl::getArrayBounds<ExecutionPolicy::array_size>(
-                begin, end, struct_bounds, s );
-            if ( a < array_bounds.second )
-                functor(
-                    Impl::Index<ExecutionPolicy::array_size>::i(s,a) );
-        };
-
-    // Execute the functor.
-    Kokkos::parallel_for( str, k_policy, functor_wrapper );
+            auto s = team.league_rank() + exec_policy.structBegin();
```

Comment:
I wonder if use `kokkos_policy team_policy( 1, exec_policy.numStruct(), VectorLength );` instead of `kokkos_policy team_policy( exec_policy.numStruct(), 1, VectorLength );`, we can just use `auto s = tream.team_rank ();`?

Reply:
have discussed this. For now what we have is functional and should get reasonable performance. We can do these explorations in future performance studies and it will not affect the API